### PR TITLE
fix: handle `IntrinsicElementalFunction_t` in `WHERE` clause

### LINKED ISSUE #4318: Bug : intrinsics applied to array inside `where` raises error after applying 'array_op' asr_pass
**MRE**
```Fortran
program name
    implicit none
    integer :: a(3)
    integer  :: b(3)
    a = [1,0,-1]
    b = 0
    where(abs(a) == 1) b = 555
    print *, b       ! Expected Result : 555 0 555  
end program name

```
Error

```bash
ASR verify pass error: The variable in ArrayItem must be an array, not a scalar
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/assem/lfortran_test/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 2481, in ??
    return main_app(argc, argv);
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 2442, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/assem/lfortran_test/lfortran/src/bin/lfortran.cpp", line 954, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/assem/lfortran_test/lfortran/src/lfortran/fortran_evaluator.cpp", line 363, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/assem/lfortran_test/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9967, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/home/assem/lfortran_test/lfortran/src/libasr/pass/pass_manager.h", line 316, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
LCompilersException: Verify failed in the pass: array_op

```

Found in files :
`outer.f90` line 341
`inner.f90` line 331

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #4318 

```fortran
program name
   implicit none
   integer :: a(3)
   integer :: b(3)
   a = [1,0,-1]
   b = 0
   where(abs(a) == 1) b = 555
   print *, b
end program name
```
is transformed similar to a function call as
```fortran
program name
implicit none
integer(4) :: __1_k
integer(4), dimension(3) :: __libasr__created__var__0__array_constant_
integer(4), dimension(3) :: __libasr__created__var__1_abs
integer(4), dimension(3) :: a
integer(4), dimension(3) :: b
__1_k = lbound(__libasr__created__var__0__array_constant_, 1)
__libasr__created__var__0__array_constant_(__1_k) = 1
__1_k = __1_k + 1
__libasr__created__var__0__array_constant_(__1_k) = 0
__1_k = __1_k + 1
__libasr__created__var__0__array_constant_(__1_k) = -1
__1_k = __1_k + 1
a = __libasr__created__var__0__array_constant_
b = 0
__libasr__created__var__1_abs = Abs(a)
do __1_k = lbound(__libasr__created__var__1_abs, 1), ubound(__libasr__created__var__1_abs, 1), 1
    if (__libasr__created__var__1_abs(__1_k) == 1) then
        b(__1_k) = 555
    end if
end do
print *, b
end program name
```

### Output
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
555 0 555 
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
Pushing tests in a while.

**Comment by Pranavchiku** [comment]
Thanks for the PR! Yes, it is fine. In general I will advise you to make a habit of pushing test at the same time, so that you know if a complete test works with your changes or not and this will make the process of reviewing PRs quicker.

**Comment by kmr-srbh** [comment]
I have pushed the test.

File: src/libasr/pass/where.cpp [hunk 1]
```diff
@@ -342,6 +343,18 @@ class WhereVisitor : public PassUtils::PassVisitor<WhereVisitor>
...
             return_var_hash[h] = opt_left;
         }
+        if (ASR::is_a<ASR::IntrinsicElementalFunction_t>(*left)) {
+            ASR::IntrinsicElementalFunction_t* i_fn = ASR::down_cast<ASR::IntrinsicElementalFunction_t>(left);
+            std::string intrinsic_name = to_lower(ASRUtils::get_intrinsic_name(i_fn->m_intrinsic_id));
+            std::string var_sym_name = current_scope->get_unique_name(intrinsic_name);
+            ASR::expr_t* return_var_expr = PassUtils::create_var(1,
+               var_sym_name, i_fn->base.base.loc,
+                i_fn->m_type, al, current_scope);
+            assign_stmt = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, return_var_expr, left, nullptr));
```
Comment (Pranavchiku) [inline src/libasr/pass/where.cpp:353]: Ohh I see what you are doing here but we do this in `intrinsic_functions` pass, let's check maybe in another PR to shift `where` pass just after `intrinsic_functions`. To do this you need to reorder passes in `pass_manager.h` and then it should be handled automatically.

**Review by Pranavchiku — COMMENTED** [review]
left a comment above

File: src/libasr/pass/where.cpp (hunk 1 above)
Comment (kmr-srbh) [reply src/libasr/pass/where.cpp:353]: Thanks for the input! It works correctly, the final result depends on #4320 though as the result is a lone `logical` array.
```fortran
where (__libasr__created__var__1__integer_comp_op_res)
    do __1_t = lbound(b, 1), ubound(b, 1)
        b(__1_t) = 555
    end do
end where
```
where `__libasr__created__var__1__integer_comp_op_res` is declared as
```fortran
logical(4), dimension(3) :: __libasr__created__var__1__integer_comp_op_res
```

**Comment by kmr-srbh** [comment]
The CI will fail for now as it depends on #4320.

**Comment by certik** [comment]
What exactly is the problem here --- the `where` pass does not handle intrinsic functions? I would rather fix the pass to work with intrinsic functions too. We want to make these passes as independent as possible.

The intrinsic function pass in the future might not expand all the intrinsics, only those that the backend does not want to handle. So all the passes should work with intrinsics.

**Comment by kmr-srbh** [comment]
> the where pass does not handle intrinsic functions? 

Yes. The initial implementation at https://github.com/lfortran/lfortran/compare/67dec734a3abd672372b71714097f7f0e538918d..e5c061475041eda1d0a5f6a553f775a98a1f859c#diff-6a1614f13a5a365acd5cab8745ebfe5ebfa8a7b15e44dd3f448ae6fcf56821af had done this.

**Comment by certik** [comment]
It seems both the FunctionCall handling as well as your initial IntrinsicElementalFunction handling that was modeled after that, only handles cases like `where (f(x) == 4)`, but not `where (4 == f(x))` or `where (f(x) + 1 == 4)`, so this approach seems just a quick fix for the (perhaps) most common case, but instead we need to be able to visit any expression in `where` in a recursive manner and replace the array with an index into the array. Something like `where (x+y*f(g(z)) == z+x)` gets transformed into `x[i]+y[i]*f(g(z[i])) == z[i] + x[i]`. This will go through the IntrinsicElementalFunction as well.

Can you test this that we don't support these nested cases yet in where?

If this requires larger refactor, then let's create an issue for this, document the problem and the actual clean solution. And in the meantime we can do these quick workarounds to unblock us for SNAP, but it's a quick patch that does not fix the underlying issue.

**Review by certik — APPROVED** [review]
The general solution is this: https://github.com/lfortran/lfortran/issues/4330, until then a simple fix like in this PR is good enough to unblock us.

**Comment by kmr-srbh** [comment]
This implementation does not work even after #4320 is merged due to some reason I do not understand clearly.

We handled logical array in that PR, but the failure here is a bit uncanny. We register correctly that the output of `abs(a) == 1` is a logical array1, but during the pass the condition `!ASR::is_a<ASR::IntegerCompare_t>(*test)` is false and we do not convert the logical array generated above into the logical binary operation we had decided upon.

I am posting the ASR and generated Fortran code after the `intrinsic_function` pass here:

### ASR
<details>

```clojure
;; ASR after applying the pass: intrinsic_function
(TranslationUnit
    (SymbolTable
        1
        {
            _lcompilers_abs_i32:
                (Function
                    (SymbolTable
                        3
                        {
                            _lcompilers_abs_i32:
                                (Variable
                                    3
                                    _lcompilers_abs_i32
                                    []
                                    ReturnVar
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            x:
                                (Variable
                                    3
                                    x
                                    []
                                    In
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    _lcompilers_abs_i32
                    (FunctionType
                        [(Integer 4)]
                        (Integer 4)
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        .false.
                    )
                    []
                    [(Var 3 x)]
                    [(If
                        (IntegerCompare
                            (Var 3 x)
                            GtE
                            (IntegerConstant 0 (Integer 4))
                            (Logical 4)
                            ()
                        )
                        [(Assignment
                            (Var 3 _lcompilers_abs_i32)
                            (Var 3 x)
                            ()
                        )]
                        [(Assignment
                            (Var 3 _lcompilers_abs_i32)
                            (IntegerUnaryMinus
                                (Var 3 x)
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                    )]
                    (Var 3 _lcompilers_abs_i32)
                    Public
                    .false.
                    .false.
                    ()
                ),
            name:
                (Program
                    (SymbolTable
                        2
                        {
                            __1_k:
                                (Variable
                                    2
                                    __1_k
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __1_t:
                                (Variable
                                    2
                                    __1_t
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __1_v:
                                (Variable
                                    2
                                    __1_v
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __libasr__created__var__0__array_constant_:
                                (Variable
                                    2
                                    __libasr__created__var__0__array_constant_
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __libasr__created__var__0__elemental_func_call_res:
                                (Variable
                                    2
                                    __libasr__created__var__0__elemental_func_call_res
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __libasr__created__var__1__integer_comp_op_res:
                                (Variable
                                    2
                                    __libasr__created__var__1__integer_comp_op_res
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Logical 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            __libasr_created_scalar_auxiliary_variable:
                                (Variable
                                    2
                                    __libasr_created_scalar_auxiliary_variable
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            a:
                                (Variable
                                    2
                                    a
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            b:
                                (Variable
                                    2
                                    b
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    name
                    []
                    [(Assignment
                        (Var 2 __1_k)
                        (ArrayBound
                            (Var 2 __libasr__created__var__0__array_constant_)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        ()
                    )
                    (Assignment
                        (ArrayItem
                            (Var 2 __libasr__created__var__0__array_constant_)
                            [(()
                            (Var 2 __1_k)
                            ())]
                            (Integer 4)
                            RowMajor
                            ()
                        )
                        (IntegerConstant 1 (Integer 4))
                        ()
                    )
                    (Assignment
                        (Var 2 __1_k)
                        (IntegerBinOp
                            (Var 2 __1_k)
                            Add
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            ()
                        )
                        ()
                    )
                    (Assignment
                        (ArrayItem
                            (Var 2 __libasr__created__var__0__array_constant_)
                            [(()
                            (Var 2 __1_k)
                            ())]
                            (Integer 4)
                            RowMajor
                            ()
                        )
                        (IntegerConstant 0 (Integer 4))
                        ()
                    )
                    (Assignment
                        (Var 2 __1_k)
                        (IntegerBinOp
                            (Var 2 __1_k)
                            Add
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            ()
                        )
                        ()
                    )
                    (Assignment
                        (ArrayItem
                            (Var 2 __libasr__created__var__0__array_constant_)
                            [(()
                            (Var 2 __1_k)
                            ())]
                            (Integer 4)
                            RowMajor
                            ()
                        )
                        (IntegerConstant -1 (Integer 4))
                        ()
                    )
                    (Assignment
                        (Var 2 __1_k)
                        (IntegerBinOp
                            (Var 2 __1_k)
                            Add
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            ()
                        )
                        ()
                    )
                    (Assignment
                        (Var 2 __1_v)
                        (ArrayBound
                            (Var 2 __libasr__created__var__0__array_constant_)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        ()
                    )
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 a)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 a)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 a)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Integer 4)
                                RowMajor
                                ()
                            )
                            (ArrayItem
                                (Var 2 __libasr__created__var__0__array_constant_)
                                [(()
                                (Var 2 __1_v)
                                ())]
                                (Integer 4)
                                RowMajor
                                ()
                            )
                            ()
                        )
                        (Assignment
                            (Var 2 __1_v)
                            (IntegerBinOp
                                (Var 2 __1_v)
                                Add
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                        []
                    )
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 b)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Integer 4)
                                RowMajor
                                ()
                            )
                            (IntegerConstant 0 (Integer 4))
                            ()
                        )]
                        []
                    )
                    (Assignment
                        (Var 2 __1_v)
                        (ArrayBound
                            (ArrayPhysicalCast
                                (Var 2 a)
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Integer 4)
                                    [((IntegerConstant 1 (Integer 4))
                                    (IntegerConstant 3 (Integer 4)))]
                                    DescriptorArray
                                )
                                ()
                            )
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        ()
                    )
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 __libasr__created__var__0__elemental_func_call_res)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 __libasr__created__var__0__elemental_func_call_res)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 __libasr__created__var__0__elemental_func_call_res)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Integer 4)
                                RowMajor
                                ()
                            )
                            (FunctionCall
                                1 _lcompilers_abs_i32
                                1 _lcompilers_abs_i32
                                [((ArrayItem
                                    (Var 2 a)
                                    [(()
                                    (Var 2 __1_v)
                                    ())]
                                    (Integer 4)
                                    RowMajor
                                    ()
                                ))]
                                (Integer 4)
                                ()
                                ()
                            )
                            ()
                        )
                        (Assignment
                            (Var 2 __1_v)
                            (IntegerBinOp
                                (Var 2 __1_v)
                                Add
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                        []
                    )
                    (Assignment
                        (Var 2 __libasr_created_scalar_auxiliary_variable)
                        (IntegerConstant 1 (Integer 4))
                        ()
                    )
                    (Assignment
                        (Var 2 __1_v)
                        (ArrayBound
                            (Var 2 __libasr__created__var__0__elemental_func_call_res)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        ()
                    )
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 __libasr__created__var__1__integer_comp_op_res)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 __libasr__created__var__1__integer_comp_op_res)
                            (IntegerConstant 1 (Integer 4))
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 __libasr__created__var__1__integer_comp_op_res)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Logical 4)
                                RowMajor
                                ()
                            )
                            (IntegerCompare
                                (ArrayItem
                                    (Var 2 __libasr__created__var__0__elemental_func_call_res)
                                    [(()
                                    (Var 2 __1_v)
                                    ())]
                                    (Integer 4)
                                    RowMajor
                                    ()
                                )
                                Eq
                                (Var 2 __libasr_created_scalar_auxiliary_variable)
                                (Logical 4)
                                ()
                            )
                            ()
                        )
                        (Assignment
                            (Var 2 __1_v)
                            (IntegerBinOp
                                (Var 2 __1_v)
                                Add
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                        []
                    )
                    (Where
                        (Var 2 __libasr__created__var__1__integer_comp_op_res)
                        [(DoLoop
                            ()
                            ((Var 2 __1_t)
                            (ArrayBound
                                (Var 2 b)
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                LBound
                                ()
                            )
                            (ArrayBound
                                (Var 2 b)
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                UBound
                                ()
                            )
                            ())
                            [(Assignment
                                (ArrayItem
                                    (Var 2 b)
                                    [(()
                                    (Var 2 __1_t)
                                    ())]
                                    (Integer 4)
                                    RowMajor
                                    ()
                                )
                                (IntegerConstant 555 (Integer 4))
                                ()
                            )]
                            []
                        )]
                        []
                    )
                    (Print
                        [(Var 2 b)]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

</details>

### Fortran code after ASR pass `intrinsic_function`

<details>

```fortran
! Fortran code after applying the pass: intrinsic_function
program name
implicit none
integer(4) :: __1_k
integer(4) :: __1_t
integer(4) :: __1_v
integer(4), dimension(3) :: __libasr__created__var__0__array_constant_
integer(4), dimension(3) :: __libasr__created__var__0__elemental_func_call_res
logical(4), dimension(3) :: __libasr__created__var__1__integer_comp_op_res
integer(4) :: __libasr_created_scalar_auxiliary_variable
integer(4), dimension(3) :: a
integer(4), dimension(3) :: b
__1_k = lbound(__libasr__created__var__0__array_constant_, 1)
__libasr__created__var__0__array_constant_(__1_k) = 1
__1_k = __1_k + 1
__libasr__created__var__0__array_constant_(__1_k) = 0
__1_k = __1_k + 1
__libasr__created__var__0__array_constant_(__1_k) = -1
__1_k = __1_k + 1
__1_v = lbound(__libasr__created__var__0__array_constant_, 1)
do __1_t = lbound(a, 1), ubound(a, 1)
    a(__1_t) = __libasr__created__var__0__array_constant_(__1_v)
    __1_v = __1_v + 1
end do
do __1_t = lbound(b, 1), ubound(b, 1)
    b(__1_t) = 0
end do
__1_v = lbound(a, 1)
do __1_t = lbound(__libasr__created__var__0__elemental_func_call_res, 1),&
         ubound(__libasr__created__var__0__elemental_func_call_res, 1)
    __libasr__created__var__0__elemental_func_call_res(__1_t) = _lcompilers_abs_i32(a(__1_v))
    __1_v = __1_v + 1
end do
__libasr_created_scalar_auxiliary_variable = 1
__1_v = lbound(__libasr__created__var__0__elemental_func_call_res, 1)
do __1_t = lbound(__libasr__created__var__1__integer_comp_op_res, 1),&
         ubound(__libasr__created__var__1__integer_comp_op_res, 1)
    __libasr__created__var__1__integer_comp_op_res(__1_t) = __libasr__created__var__0__elemental_func_call_res(__1_v) ==&
         __libasr_created_scalar_auxiliary_variable
    __1_v = __1_v + 1
end do
where (__libasr__created__var__1__integer_comp_op_res)
    do __1_t = lbound(b, 1), ubound(b, 1)
        b(__1_t) = 555
    end do
end where
print *, b

contains

integer(4) function _lcompilers_abs_i32(x)
    integer(4), intent(in) :: x
    if (x >= 0) then
        _lcompilers_abs_i32 = x
    else
        _lcompilers_abs_i32 = -x
    end if
end function _lcompilers_abs_i32

end program name
```

</details>

@Pranavchiku if you have time could you please help me with this? I had tried to execute program given in #4317 and it had a similar code generated after this pass, it fails there too, but I am not sure if it will fix it.

**Comment by Pranavchiku** [comment]
Yes, I'll check it in a while.

**Comment by Pranavchiku** [comment]
This diff against latest main, makes the code running, although output is `555 555 555`, I suspect something is off which I will leave upto you to figure out and eventually fix:


```diff
diff --git a/src/libasr/pass/pass_manager.h b/src/libasr/pass/pass_manager.h
index 6cd546f36..31516c55d 100644
--- a/src/libasr/pass/pass_manager.h
+++ b/src/libasr/pass/pass_manager.h
@@ -216,13 +216,13 @@ namespace LCompilers {
                 "implied_do_loops",
                 "class_constructor",
                 "pass_list_expr",
-                "where",
                 "function_call_in_declaration",
                 "subroutine_from_function",
                 "array_op",
                 "symbolic",
                 "intrinsic_function",
                 "intrinsic_subroutine",
+                "where",
                 "subroutine_from_function",
                 "array_op",
                 "pass_array_by_data",
@@ -250,7 +250,6 @@ namespace LCompilers {
                 "implied_do_loops",
                 "class_constructor",
                 "pass_list_expr",
-                "where",
                 "function_call_in_declaration",
                 "subroutine_from_function",
                 "array_op",
@@ -258,6 +257,7 @@ namespace LCompilers {
                 "flip_sign",
                 "intrinsic_function",
                 "intrinsic_subroutine",
+                "where",
                 "subroutine_from_function",
                 "array_op",
                 "pass_array_by_data",
diff --git a/src/libasr/pass/where.cpp b/src/libasr/pass/where.cpp
index 312deabcb..c1c40eb94 100644
--- a/src/libasr/pass/where.cpp
+++ b/src/libasr/pass/where.cpp
@@ -2,6 +2,7 @@
 #include <libasr/containers.h>
 #include <libasr/exception.h>
 #include <libasr/asr_utils.h>
+#include <libasr/asr_builder.h>
 #include <libasr/asr_verify.h>
 #include <libasr/pass/pass_utils.h>
 #include <libasr/pass/replace_where.h>
@@ -305,6 +306,18 @@ public:
         ASR::stmt_t* assign_stmt = nullptr;
         ASR::stmt_t* if_stmt = nullptr;
 
+        std::cout<<"ASRUtils::type_to_str_python(ASRUtils::expr_type(test)): "<<ASRUtils::type_to_str_python(ASRUtils::expr_type(test))<<std::endl;
+        if (ASR::is_a<ASR::Var_t>(*test) && ASRUtils::is_logical(*ASRUtils::expr_type(test))) {
+            std::cout<<"in here"<<std::endl;
+            ASR::Var_t* var = ASR::down_cast<ASR::Var_t>(test);
+            ASRUtils::ASRBuilder b(al, var->base.base.loc);
+            ASR::expr_t* logical_true = ASRUtils::EXPR(
+                                            ASR::make_LogicalConstant_t(al, x.base.base.loc, true,
+                                            ASRUtils::TYPE(ASR::make_Logical_t(al, x.base.base.loc, 4))));
+            test = ASRUtils::EXPR(ASR::make_LogicalBinOp_t(al, x.base.base.loc, test,
+                        ASR::logicalbinopType::Eqv, logical_true, ASRUtils::expr_type(test), nullptr));
+        }
+ 
         if (ASR::is_a<ASR::IntegerCompare_t>(*test)) {
             int_cmp = ASR::down_cast<ASR::IntegerCompare_t>(test);
             left = int_cmp->m_left;
```

```fortra
program name
   implicit none
   integer :: a(3)
   integer :: b(3)
   a = [1,0,-1]
   b = 0
   where(abs(a) == 1) b = 555
   print *, b
end program name
```

```console
% gfortran a.f90 && ./a.out
         555           0         555
% lfortran a.f90 
555 555 555 
```

**Review by Pranavchiku — COMMENTED** [review]
left a comment above

**Comment by Pranavchiku** [comment]
When we look at translation of `where`, it is incorrect:

```fortran
do __1_k = lbound(__libasr__created__var__1__integer_comp_op_res, 1),&
         ubound(__libasr__created__var__1__integer_comp_op_res, 1), 1
    if (__libasr__created__var__1__integer_comp_op_res(__1_k) .eqv. .true.) then
        do __1_t = lbound(b, 1), ubound(b, 1)
            b(__1_t) = 555
        end do
    end if
end do
```

It should be:

```fortran
do __1_k = lbound(__libasr__created__var__1__integer_comp_op_res, 1),&
         ubound(__libasr__created__var__1__integer_comp_op_res, 1), 1
    if (__libasr__created__var__1__integer_comp_op_res(__1_k) .eqv. .true.) then
            b(__1_k) = 555
    end if
end do
```

and this happens in pass `array_op` which means we need to have `where` before `array_op`.

**Comment by kmr-srbh** [comment]
> This diff against latest main, makes the code running

This diff looks good. I will try to get this working. Thanks for looking into this!

**Comment by Pranavchiku** [comment]
I opened PR at https://github.com/lfortran/lfortran/pull/4360, you may feel free to continue that.

**Comment by Pranavchiku** [comment]
Ahh I had a comment that stated:

we have to use the following diff against latest main:

```diff
diff --git a/src/libasr/pass/where.cpp b/src/libasr/pass/where.cpp
index 312deabcb..572bfb426 100644
--- a/src/libasr/pass/where.cpp
+++ b/src/libasr/pass/where.cpp
@@ -2,6 +2,7 @@
 #include <libasr/containers.h>
 #include <libasr/exception.h>
 #include <libasr/asr_utils.h>
+#include <libasr/asr_builder.h>
 #include <libasr/asr_verify.h>
 #include <libasr/pass/pass_utils.h>
 #include <libasr/pass/replace_where.h>
@@ -343,25 +344,26 @@ public:
             return_var_hash[h] = opt_left;
         }
 
+        ASRUtils::ExprStmtDuplicator dup(al);
         if (opt_left && ASR::is_a<ASR::IntegerCompare_t>(*test)) {
-            int_cmp = ASR::down_cast<ASR::IntegerCompare_t>(test);
-            int_cmp->m_left = opt_left;
+            int_cmp = ASR::down_cast<ASR::IntegerCompare_t>(dup.duplicate_expr(test));
+            int_cmp->m_left = dup.duplicate_expr(opt_left);
         }
         if (opt_left && ASR::is_a<ASR::RealCompare_t>(*test)) {
-            real_cmp = ASR::down_cast<ASR::RealCompare_t>(test);
-            real_cmp->m_left = opt_left;
+            real_cmp = ASR::down_cast<ASR::RealCompare_t>(dup.duplicate_expr(test));
+            real_cmp->m_left = dup.duplicate_expr(opt_left);
         }
         if (opt_left && ASR::is_a<ASR::LogicalBinOp_t>(*test)) {
-            log_bin_op = ASR::down_cast<ASR::LogicalBinOp_t>(test);
-            log_bin_op->m_left = opt_left;
+            log_bin_op = ASR::down_cast<ASR::LogicalBinOp_t>(dup.duplicate_expr(test));
+            log_bin_op->m_left = dup.duplicate_expr(opt_left);
         }
 
         // create do loop head
         ASR::do_loop_head_t head;
         head.loc = loc;
         head.m_v = var;
-        head.m_start = PassUtils::get_bound(opt_left?opt_left:left, 1, "lbound", al);
-        head.m_end = PassUtils::get_bound(opt_left?opt_left:left, 1, "ubound", al);
+        head.m_start = PassUtils::get_bound(opt_left?dup.duplicate_expr(opt_left):dup.duplicate_expr(left), 1, "lbound", al);
+        head.m_end = PassUtils::get_bound(opt_left?dup.duplicate_expr(opt_left):dup.duplicate_expr(left), 1, "ubound", al);
         head.m_increment = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int32_type));
 
         // create do loop body
```

The trouble we had was that `left` expression had same pointer and thus one change in `array_op` updated it at all places and then consequent steps were wrong. So we have to duplicate expr to avoid it.

**Comment by kmr-srbh** [comment]
The new changes with #4360 fixes the issue.

```fortran
program name
   implicit none
   integer :: a(3)
   integer :: b(3)
   a = [1,0,-1]
   b = 0
   where(abs(a) == 1) b = 555
   print *, b
end program name
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
555 0 555 
```

@Pranavchiku shall we close this PR?

**Pull request closed without merging**